usb: allow hardware to handle ZLP for Variable-length Data Stage

jfischer-no opened the pull request:
Allow hardware to handle ZLP for Variable-length Data Stage.

On nRF52840, USB hardware and driver [handle](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/usb/device/usb_dc_nrfx.c#L1679) the status stage. When it comes to Variable-length Data Stage nRF52 hardware finishes a transmission with ZLP.

How to test:
Apply the patch below:
```diff
diff --git a/subsys/usb/class/loopback.c b/subsys/usb/class/loopback.c
index 66e2aaf0b8..49f37709f1 100644
--- a/subsys/usb/class/loopback.c
+++ b/subsys/usb/class/loopback.c
@@ -144,6 +144,9 @@ static int loopback_vendor_handler(struct usb_setup_packet *setup,
            (setup->bRequest == 0x5c)) {
                LOG_DBG("Device-to-Host, wLength %d, data %p",
                        setup->wLength, *data);
+               if (setup->wLength == 1024) {
+                       *len = 512;
+               }
                return 0;
        }

```
Perform TEST14: e.g. `testusb -v 1 -t 14 -c 1000 -D /dev/bus/usb/005/056`
on the current master nRF52840 device will not able to transmit on the control IN endpoint after first 512 bytes. With this patch, testusb reports `Broken pipe` but device remains full functional.

Alternatively, `cdc_acm_composite` sample can be used as it has a string descriptor exact 64 bytes long.

rebased on #19321

jfischer-no committed:
usb: allow hardware to handle ZLP for Variable-length Data Stage

Allow hardware to handle ZLP for Variable-length Data Stage.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

carlescufi commented on subsys/usb/Kconfig:
```diff
@@ -79,6 +79,12 @@ config USB_DEVICE_REMOTE_WAKEUP
 	help
 	  This option requires USBD peripheral driver to also support remote wakeup.
 
+config USB_DEVICE_DRIVER_HANDLE_CTRL_EPIN_ZLP
```
Shouldn't you be enabling this by default in Nordic ICs?

jfischer-phytec-iot replied:
:man_facepalming: forgot `git add drivers`

emob-nordic commented on samples/subsys/usb/cdc_acm_composite/prj.conf:
```diff
@@ -12,6 +12,7 @@ CONFIG_USB_COMPOSITE_DEVICE=y
 CONFIG_USB_CDC_ACM=y
 CONFIG_USB_CDC_ACM_DEVICE_COUNT=2
 CONFIG_USB_CDC_ACM_RINGBUF_SIZE=512
+CONFIG_USB_REQUEST_BUFFER_SIZE=192
```
I think its not related to this PR and should be removed or am I missing something?

jfischer-phytec-iot replied:
It is based on #19203 and there is no problem with that.

emob-nordic replied:
Its not a problem with a change itself. This PR is about ZLP handling so I would avoid putting this change here as it is not related.

jfischer-phytec-iot replied:
It is related because "Alternatively, cdc_acm_composite sample can be used as it has a string descriptor exact 64 bytes long." To do that #19203 is necessary.

emob-nordic commented on samples/subsys/usb/hid-cdc/prj.conf:
```diff
@@ -14,6 +14,7 @@ CONFIG_USB_DEVICE_LOG_LEVEL_INF=y
 
 CONFIG_USB_CDC_ACM=y
 CONFIG_USB_CDC_ACM_DEVICE_COUNT=2
+CONFIG_USB_REQUEST_BUFFER_SIZE=192
```
I think its not related to this PR and should be removed or am I missing something?

emob-nordic commented on samples/subsys/usb/hid-cdc/src/main.c:
```diff
@@ -630,6 +630,9 @@ void main(void)
 		uart_line_ctrl_get(cdc0_dev, LINE_CTRL_DTR, &dtr);
 		if (dtr) {
 			break;
+		} else {
```
I think for this another PR is already pending and its not related to ZLP.

emob-nordic commented on samples/subsys/usb/hid-cdc/src/main.c:
```diff
@@ -639,6 +642,9 @@ void main(void)
 		uart_line_ctrl_get(cdc1_dev, LINE_CTRL_DTR, &dtr);
 		if (dtr) {
 			break;
+		} else {
```
Same here.

emob-nordic commented on subsys/usb/Kconfig:
```diff
@@ -79,6 +79,12 @@ config USB_DEVICE_REMOTE_WAKEUP
+	bool
+	help
+	  Stack should not handle ZLP for Variable-length Data Stage
+	  bacause it is taken over by the hardware.
```
Not by hardware but by LL drivers I think (for nRF for sure).

jfischer-phytec-iot replied:
There is comment in the nRF driver:  "a flag is set to prevent triggering status stage which is handled by hardware". If **Not by hardware but by LL drivers I think (for nRF for sure)** is true then please fix your driver so we do not have to do something ugly in the stack!

emob-nordic commented on subsys/usb/Kconfig:
```diff
@@ -79,6 +79,12 @@ config USB_DEVICE_REMOTE_WAKEUP
 	help
 	  This option requires USBD peripheral driver to also support remote wakeup.
 
+config USB_DEVICE_DRIVER_HANDLE_CTRL_EPIN_ZLP
```
I would rather go for a name like:
USB_DEVICE_DISABLE_STACK_ZLP or similar to explicitly show that this option disable ZLP handling in zephyr usb stack.

jfischer-phytec-iot replied:
No, because it does not describe why.

carlescufi replied:
`USB_DEVICE_ZLP_EPIN_HW` or `USB_DEVICE_DISABLE_ZLP_EPIN_HANDLING`

jfischer-phytec-iot replied:
changed to USB_DEVICE_DISABLE_ZLP_EPIN_HANDLING

emob-nordic commented on subsys/usb/usb_device.c:
```diff
@@ -332,6 +334,11 @@ static void usb_handle_control_transfer(u8_t ep,
 
 		/* Send smallest of requested and offered length */
 		usb_dev.data_buf_residue = MIN(usb_dev.data_buf_len, length);
+		if (usb_dev.data_buf_len > CONFIG_USB_REQUEST_BUFFER_SIZE) {
```
This is part of another PR and should not be there. It's not mentioned in PR message.

emob-nordic commented on drivers/usb/device/Kconfig:
```diff
@@ -61,6 +61,7 @@ menuconfig USB_NRF52840
 	depends on SOC_NRF52840
 	select USB_DEVICE_DRIVER
 	select NRFX_USBD
+	select USB_DEVICE_DRIVER_HANDLE_CTRL_EPIN_ZLP
```
Name should be different.
Check **subsys/usb/Kconfig** comment.

eobalski reviewed (changes requested):
Only this [https://github.com/zephyrproject-rtos/zephyr/pull/19272/commits/afec959e1d91bf57f7daf068af959f40d282057e](url) commit is actually related to PR message IMO.

I would change the name of the PR for something like:
**"usb: disable ZLP handling by zephyr usb stack for variable-length data stage"**
as zlp is not handled by hw only but by LL driver.

jfischer-no committed:
usb: do not check size of REQUEST_BUFFER for Data stage IN

Do not check size of REQUEST_BUFFER for Data stage IN
transactions. The check can not be done effectively because
the pointer (usb_dev.data_buf) can be changed and the actual
size of the buffer used is unknown at this point.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-phytec-iot commented:
@emob-nordic @carlescufi 
This ZLP issue can be reproduced with cdc_acm_composite sample (patch 52f040c4e8a6b68be4cb1bf0e098087003cbaffb is necessary because of request length check issue). The logs should look something like this:

��***** Booting Zephyr OS build zephyr-v2.0.0-633-g52f040c4e8a6 *****
[00:00:00.005,157] <inf> cdc_acm_composite: Wait for DTR
[00:00:00.009,185] <inf> usb_cdc_acm: USB device suspended
[00:00:00.009,216] <inf> usb_cdc_acm: USB device suspended
[00:00:00.149,291] <inf> usb_cdc_acm: USB device resumed
[00:00:00.149,322] <inf> usb_cdc_acm: USB device resumed
[00:00:00.220,489] <inf> usb_cdc_acm: USB device suspended
[00:00:00.220,489] <inf> usb_cdc_acm: USB device suspended
[00:00:00.389,251] <inf> usb_cdc_acm: USB device resumed
[00:00:00.389,251] <inf> usb_cdc_acm: USB device resumed
[00:00:00.705,810] <err> usb_nrfx: Unexpected event (nrfx_usbd): 3, ep 0x80
[00:00:31.317,871] <inf> usb_cdc_acm: USB device configured
[00:00:31.317,871] <inf> usb_cdc_acm: USB device configured
